### PR TITLE
feat: add location option support and trino dialect

### PR DESCRIPTION
**Pull request by qwinzIdo**
Partially targeting https://github.com/taozhi8833998/node-sql-parser/issues/316
This is a suggested addition to support node location objects in the parsed AST.
This change introduces the `includeLocations` option to the `astify()` method - when  set to `true`, a `loc` property will be added the the parsed nodes.
The peg implementation is done by adding `...getLocationObject()` in the returned object of the desired peg rule. `getLocationObject()` checks for the `includeLocations` option and returns a `{ loc: location() }` object.
While we've added the `getLocationObject()` function in all supported dialects, we only added the call to some rules in `mysql.pegjs`: 
1. This is what we initially need, but we will need more, from more dialects, soon
2. As adding full support is a pretty massive change, we think this can be done in iterations
3. We first wanted to get an initial approval for the suggested implementation method

Opinions, suggestion or complete rejections are welcomed :)

**Update:**
This PR was hijacked to also include Trino SQL dialect support :)

### CODE REVIEW

**Comment by qwinzIdo** [comment]
@taozhi8833998 added partial bigquery location indication as well

**Comment by qwinzIdo** [comment]
@taozhi8833998 Added `pegjs` to prod dependencies for its `Location` and `LocationRange` types and made adjustments for the latest bigquery fix you just introduced

**Comment by darzIdo** [comment]
@taozhi8833998 Added a `trimQuery` option (defaults to `true` to keep the current default behavior), to allow the user to get node locations with true offsets

**Comment by darzIdo** [comment]
@taozhi8833998 introduced snowflake support

**Comment by taozhi8833998** [comment]
@qwinzIdo @darzIdo Thanks for the pr, reviewing it.

**Comment by taozhi8833998** [comment]
releated to #1457

**Comment by taozhi8833998** [comment]
@darzIdo Can you add me as your fork repository maintainer? so I can push codes to the [repo](git@github.com:darz-io/node-sql-parser.git) for lambda expressions in trino or I open a new pr, and include current pr?

**Comment by taozhi8833998** [comment]
image

**Comment by darzIdo** [comment]
@taozhi8833998 just added you now

**Comment by taozhi8833998** [comment]
@darzIdo ok, push the codes that support lambda expressions in trino.

**Comment by darzIdo** [comment]
@taozhi8833998 lambda confirmed to be working :) thanks!!
I have a couple more minor issues to solve, i'll fix them and ping you when this can be merged

**Comment by taozhi8833998** [comment]
@darzIdo ok, thanks for the great job. If any new features I can add, please let me know.

**Comment by darzIdo** [comment]
@taozhi8833998 I do see one thing regarding `...rows between` clause, which i think is not special to trino and might affect all dialects.
trying to parse the following statement fails:
`select sum(a) over (partition by b rows between unbounded preceding and current row)`

while this one is parsed as expected (`order` instead of `partition`)
`select sum(a) over (order by b rows between unbounded preceding and current row)`

I'm working on fixing it, but if this is something you can quickly spot the solution to, it would be great :)

**Comment by taozhi8833998** [comment]
@darzIdo In trino, the function name can be quoted or not? such as **"funcName"(a, b, c)**

**Comment by darzIdo** [comment]
@taozhi8833998, yes - seems like it can be quoted

**Comment by taozhi8833998** [comment]
@darzIdo If it can be quoted, we should using object to keep the function name original quoted symbol when sqlify back.

**Comment by darzIdo** [comment]
Got it, so the types should be changed I guess. How about instead of changing the type of the `name` field, we'll add a new `nameOrigin` field or something similar? To keep it consistent with the rest of the dialects.
Same goes for the column right? I saw you change that as well...

And did you get a chance to take a llok at the "partition by" issue I described above?

**Comment by taozhi8833998** [comment]
> Got it, so the types should be changed I guess. How about instead of changing the type of the `name` field, we'll add a new `nameOrigin` field or something similar? To keep it consistent with the rest of the dialects. Same goes for the column right? I saw you change that as well...
> 
> And did you get a chance to take a llok at the "partition by" issue I described above?

@darzIdo Sorry I missed the message, I will check it now.

**Comment by taozhi8833998** [comment]
@darzIdo fixed the "partition by" issue

**Comment by darzIdo** [comment]
@taozhi8833998 great thanks! And the `name` thing I asked above, what do you say? Should we foze the typing? Or add a new field?

**Comment by taozhi8833998** [comment]
Is it complex to change the typing? I will change the **name** field in all dialects to keep the same type.

**Comment by darzIdo** [comment]
@taozhi8833998 no, it's very simple - just tell me what fields have changed so I won't miss anything and I can make the change in type.d.ts.
One important thing to notice is that it's a breaking change, so we'll have to bump the major version to 5.x.x

**Comment by taozhi8833998** [comment]
@darzIdo ok, The next version will be 5.0.0

**Comment by darzIdo** [comment]
@taozhi8833998 no problem, just ping when you made the change, I'll update the typings

**Comment by darzIdo** [comment]
>@taozhi8833998 no problem, just ping when you made the change, I'll update the typings

@taozhi8833998 any update on this? thanks 🙏

**Comment by taozhi8833998** [comment]
> > @taozhi8833998 no problem, just ping when you made the change, I'll update the typings
> 
> @taozhi8833998 any update on this? thanks 🙏

 fixed in #1827

**Comment by darzIdo** [comment]
@taozhi8833998 trino dialect updated with the latest function name changes. I think we can merge :)

**Comment by darzIdo** [comment]
@taozhi8833998 updated again with types after merging https://github.com/taozhi8833998/node-sql-parser/pull/1831

**Comment by darzIdo** [comment]
@taozhi8833998 can this be merged?

**Comment by taozhi8833998** [comment]
@darzIdo I am busy for this days. Sorry for the delay. I will merge it this weekend,  and publish `5.0.0` with your hard working. Thanks for the great pr.

**Review by taozhi8833998 — APPROVED** [review]
agree

**Pull request merged**
